### PR TITLE
EE-2575 - Improved JSON validation GH actions

### PR DESCRIPTION
**Pull request by M-casado**
## Ticket reference
EE-2575

## Overall changes
- Changed JSON documents to contain both the ``schema`` and the ``data`` that Biovalidator expects (see [documentation](https://github.com/elixir-europe/biovalidator#using-biovalidator-as-a-server)). Previously they only contained the ``data``. 
- Removed previous JSON validation GitHub action (``json_validation.yml``)
- Added new GH action to validate ``branch`` JSON documents against EGA's API (with remote ``main`` schemas)
- Added new GH action to validate ``branch`` JSON documents against a locally deployed Biovalidator (with ``branch`` schemas)
- Added a script (``.github/scripts/request_validation.py``) that is used by both actions to do POST requests and interpret the results.

## Future TO-DOs
- [ ] Keep pushing for https://github.com/elixir-europe/biovalidator/issues/53 to be amended (otherwise the validation will never be passed)
- [x] After reviewed, update "https://github.com/EbiEga/ega-metadata-schema/tree/main/docs/gh_workflows" with the new GitHub actions and removing the previous one.

## Special focus for reviewers
I would focus on the logic behind the two new actions, described in the documentation and comments of both, and the Python script.

Keep in mind that the GH actions do work: I have tested both of them passing and failing validation (check [this PR](https://github.com/M-casado/ega-metadata-schema/pull/1)), but they won't pass validation in this case until https://github.com/elixir-europe/biovalidator/issues/53 is resolved.

### CODE REVIEW

File: .github/workflows/json_validation_deploying_biovalidator.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+---
```
Comment (aaclan-ebi) [inline .github/workflows/json_validation_deploying_biovalidator.yml:1]: i think this is out of scope of this repo

**Review by aaclan-ebi — COMMENTED** [review]
Hi Marcos, Nice work! I've got some suggestions. Hope you find them useful.

File: .github/scripts/request_validation.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+extension = ".json"
+
+# --------- #
+# Function definition
+# --------- #
+
+
+def request_validation(data_filepath, 
+                       curl_URL,
+                       headers={'Content-Type': 'application/json'}
```
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py:21]: Amended 👍

File: .github/workflows/json_validation_deploying_biovalidator.yml (hunk 1 above)
Comment (M-casado) [reply .github/workflows/json_validation_deploying_biovalidator.yml:1]: Sorry can you elaborate on this one, Alegria? Out of the scope of the repo the whole YML file, the GH action itself?

File: .github/scripts/request_validation.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    with open(data_filepath) as f:
+        data = f.read().replace('\n', '').replace('\r', '').encode('utf-8')
+
+    response = requests.post(url=curl_URL, headers=headers, data=data)
+    return response
+
+
+def get_errors_response(response):
+    """
+    Function that, given a "requests.models.Response" object, will interpret it, asserting that the request
```
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py:36]: Added the type of every arg. coming in and out of the functions 👍

File: .github/scripts/request_validation.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        was successful and no validation errors (i.e. empty list of errors) were found. If not, it will
+        return the error.
+    """
+    filename = file.name
+    assert (
+            type(response) == requests.models.Response
+        ), "The POST response was not of the correct type"
+    
+    if not response.status_code == 200:
```
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py:45]: Done

File: .github/scripts/request_validation.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+for file in os.scandir(dirname):
+    # Check if it's a JSON file
+    if not file.path.endswith(extension):
+        continue
+
+    request = request_validation(data_filepath=file,
+                                 curl_URL=curl_URL)
+
+    val_error = get_errors_response(request)
+    # If there was a validation error, we add it to the dictionary
```
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py:71]: Well, personally I prefer one-liners that are easy to interpret for those who have not read the code. Since they don't influence the execution time and can be ignored, or even parsed out, if needed, I tend to have a few in my code. 

Once again, it's personal preference, since I've had to read my own code after a year and I appreciated these. Or would have paid for them to be present in many other scripts. 

In any case, I left only 3 comments that, to my taste, add context.

File: .github/scripts/request_validation.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    if val_error:
+        error_dict[file.name] = val_error
+
+# If the dictionary is not empty, fail the script returning it
+if error_dict:
+    sys.exit(error_dict)
```
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py:77]: Taken from latest [documentation](https://docs.python.org/3/library/sys.html), using ``arg`` as intended, I believe:
image

Passing the error message as ``arg`` to ``sys.exit()`` will result in the exit code being ``1``, and having the error message given as ``stderr``. This is how it gets displayed in the GitHub action:
image

File: .github/workflows/json_validation_deploying_biovalidator.yml (hunk 1 above)
Comment (aaclan-ebi) [reply .github/workflows/json_validation_deploying_biovalidator.yml:1]: I think testing if the current "main" branch of Biovalidator can be deployed as the instance is out of the scope of this repo. That test should be in the Biovalidator repo itself.

**Comment by M-casado** [comment]
I just noticed that the installation of Biovalidator is failing now, I'll contact their maintainers to check on what changed for this to pop-up now.

File: .github/scripts/request_validation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaclan-ebi) [reply .github/scripts/request_validation.py]: Comments can distract you from reading the code, try to focus more on the readability of the code than putting comments to explain them. I really like it when it feels like you're reading an English text/story when you're reading the code. Using descriptive names and functions can help. You don't have to apply it now only if time permits and yeah just think about it next time when you're writing/refactoring your codes.

for example(may not be the best example but just trying to make a point that you don't need the comments in some parts of your code):

Before:
```
# We iterate over the JSON documents to validate
for file in os.scandir(dirname):
    # Check if it's a JSON file
    if not file.path.endswith(extension):
        continue
```


After:
```
for file in os.scandir(dirname):
    if not is_json(file):
        continue

def is_json(file):
    return file.path.endswith(extension)
```

File: .github/scripts/request_validation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaclan-ebi) [reply .github/scripts/request_validation.py:76]: I guess that's fine. Oh, you could also improve the printing of the errors by doing:

```
json.dumps(error_dict, sort_keys=True, indent=4)
```

File: .github/workflows/json_validation_deploying_biovalidator.yml (hunk 1 above)
Comment (M-casado) [reply .github/workflows/json_validation_deploying_biovalidator.yml:1]: Even if we don't need it here, giving the schemas is done (at least with how Biovalidator works for now) at the deployment level. We **need** to deploy it and give the references ``-r`` with the branch's schemas. 
The most we could perhaps do is have an installation in some GH space and deploy it with different schemas each time the GH actions are executed. But I feel like that's a hassle and would be just easier to install it at every push. 

For normal PR I tend to push only after a few commits are added to it, so I wouldn't say the GH action would be executed that many times. Besides, it normally takes less than 30 seconds (~20-25 secs I think).

**Review by aaclan-ebi — APPROVED** [review]
Thanks for considering and applying the suggestions 👍

File: .github/scripts/request_validation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (M-casado) [reply .github/scripts/request_validation.py]: Yep, understood. Will try to keep it in mind, although I tend to be... verbose. Thanks for the feedback, Alegria 🥇

**Comment by M-casado** [comment]
Addressed all of the complains from the SuperLinter, ready to merge
